Add yaml linting

viccuad opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

This PR adds yaml linting to the dev/ scripts and gitlab, which lints for yaml files, excluding `deploy/helm/kubecf/*`, as those are helm templates that should be better linted with `helm lint`.

To achieve it, it adds a bazel target `//dev/linters:yamllint` to install the yamllint python module, and a python wrapper to call it. I haven't found any other way in Bazel to pull a python module and then call it.

It provides the script `dev/linters/yamllint.sh` that calls the python wrapper. And it updates `.gitlab-ci.yaml` to call this last script.

Bazel's pip_install() doesn't install pip, so this PR needs https://github.com/f0rmiga/bazel-docker-image/pull/3.


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- An important detail to include is the version of the used cf-operator -->

I have tested it locally by running `dev/linters/yamllint.sh`.

I have tested it against gitlab-ci by editing the `.gitlab-ci.yaml` to `zypper in pip`, and running `gitlab-runner exec docker lint`.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code has security implications.
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.

viccuad commented:
Linting the helm templates fail, eg:
```
./deploy/helm/kubecf/assets/operations/instance_groups/bits.yaml
  1:3       error    syntax error: expected the node content, but found '-'
```
We can't use `helm lint` as it only accepts `Chart.yml` as input. Another option would be to litter all templates with `# yamllint disable-line`, but that doesn't seem like a better option.

f0rmiga commented:
https://github.com/SUSE/kubecf/pull/165 Should unblock this PR.

viccuad commented:
Rebased on top of master to include https://github.com/SUSE/kubecf/pull/165.
Edited PR message to signify that this PR lints yaml files excluding `deploy/helm/kubecf/*`, as those are helm templates that should be better linted with `helm lint`

f0rmiga commented on WORKSPACE:
```diff
@@ -13,6 +13,20 @@ external_binary(
 
+http_archive(
+    name = "rules_python",
+    url = "https://github.com/bazelbuild/rules_python/releases/download/0.0.1/rules_python-0.0.1.tar.gz",
```
We are controlling all the repository rules versions and hash sums on the top-level def.bzl. Check the `bazel_skylib` target:

https://github.com/SUSE/kubecf/blob/6e48b3fd04c40d2a64d5faeb33b2d6f73a194d32/WORKSPACE#L73-L77

https://github.com/SUSE/kubecf/blob/6e48b3fd04c40d2a64d5faeb33b2d6f73a194d32/def.bzl#L91-L94

f0rmiga commented on WORKSPACE:
```diff
@@ -13,6 +13,20 @@ external_binary(
+    url = "https://github.com/bazelbuild/rules_python/releases/download/0.0.1/rules_python-0.0.1.tar.gz",
+    sha256 = "aa96a691d3a8177f3215b14b0edc9641787abaaa30363a080165d06ab65e1161",
+)
+load("@rules_python//python:pip.bzl", "pip_repositories","pip_import")
```
Could you add a line break between each function call, please? Once we add `skylint`, it will ask the same thing. :)

f0rmiga commented on dev/linters/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+py_binary(
+    name = "yamllint",
+    srcs = ["//dev/linters:run_yamllint.py"],
+    data = [
```
Remove `data` completely, it's not being used.

f0rmiga commented on dev/linters/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+    ],
+    main = "run_yamllint.py",
+    python_version = "PY2",
+    deps = [
```
Unnecessary `[]`. Just do `deps = all_requirements`.

f0rmiga commented on dev/linters/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+    python_version = "PY2",
+    deps = [
+    ] + all_requirements,
+)
```
Missing line break at the end of the file.

f0rmiga commented on dev/linters/run_yamllint.py:
```diff
@@ -0,0 +1,10 @@
+import os
+from yamllint import cli
+# import yamllint
```
Remove commented code.

f0rmiga commented on doc/Contribute.md:
```diff
@@ -97,18 +97,20 @@ associated documentation, if we have any.
 ## Linting
 
-Currently only one linter is available:
+Currently 2 linters are available:
```
[nit] Missing a comma after `Currently`.

f0rmiga commented on doc/Contribute.md:
```diff
@@ -129,7 +131,7 @@ around the deployment manifest the ability to easily toggle various
 as the application scheduler.
 
-### Features
+
```
Why this change?

f0rmiga commented on dev/linters/yamllint.sh:
```diff
@@ -0,0 +1,9 @@
+# We want word splitting with find.
+bazel run //dev/linters:yamllint -- \
+      -d "{extends: relaxed, rules: {line-length: {max: 120}}}" \
+      --strict $(find . -not -path "./deploy/helm/kubecf/*" -name '*.yaml' -or -name '*.yml' )
```
[nit] Missing type file only.
[Super nit] Extra space on the closing parenthesis.

f0rmiga replied:
I think we still want to lint `values.yaml`.

f0rmiga replied:
I'm wondering if it would worth running the `yamllint` on the rendered templates. It would require inverting the build step with the linting step on CI. Something like:

```
bazel run //dev/linters:yamllint -- ... <(helm template --name kubecf ...)
```

Ping @mook-as.

mook-as replied:
Running linters on generated output can often be a bad idea (because we would often get generated output that would be different than how a human would write things from scratch, e.g. with very long lines).

f0rmiga replied:
In this case, we should aim to generate human-like manifests. A linter on the generated output would help us with achieving that. It's a different case than running a linter on, let's say, gRPC generated code, although I'd argue that it may be subject to some pre-compile checks. The reason why I believe we should aim to generate human-like manifests is that it makes inspection easy for everyone from devs to final customers.

viccuad replied:
Good catch, I have added `deploy/helm/kubecf/values.*`
We could always lint before and after build.

mook-as commented on dev/linters/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+        # "//dev/linters:setup.cfg",
+    ],
+    main = "run_yamllint.py",
+    python_version = "PY2",
```
We should probably be using Python3 throughout instead? I don't think we have any other Python dependencies at the moment, and Python 2 is going to EOL "soon". (And yamllint supports Python 3, as far as I can tell.)

f0rmiga replied:
+1

Both Python 2 and 3 are available on the Docker image (along with pip). I'll drop Python 2 as soon as it reaches EOL.

viccuad commented:
Many thanks for the review :). Also, I'm quite embarrassed by the simple mistakes I did.

I have fixed the typos and force-pushed with new changes, which I think it's gonna be cleaner:
- Using `git_repository` instead of `http_archive`, needed for python3 support, saving the commit sha in `def.bzl`
- pip_repositories and pip_import need to be evaluated at WORKSPACE-time, added commit comments. The rest can happily go on ./dev/linters/BUILD.bazel.

viccuad commented:
Rebased on top of master to fix conflicts.

mook-as commented on dev/linters/yamllint.sh:
```diff
@@ -0,0 +1,12 @@
+
+# shellcheck disable=SC2046
+# We want word splitting with find.
+bazel run //dev/linters:yamllint -- \
```
I'm having issues running this; I get:

```
$ bazel run //dev/linters:yamllint
Starting local Bazel server and connecting to it...
DEBUG: Rule 'rules_python' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1573842889 -0500"
INFO: Analyzed target //dev/linters:yamllint (18 packages loaded, 210 targets configured).
INFO: Found 1 target...
INFO: Deleting stale sandbox base …/1902c1ccc22df327a938dca0d6a953d0/sandbox
Target //dev/linters:yamllint up-to-date:
  bazel-bin/dev/linters/yamllint
INFO: Elapsed time: 14.173s, Critical Path: 0.06s
INFO: 0 processes.
INFO: Build completed successfully, 4 total actions
INFO: Build completed successfully, 4 total actions
Traceback (most recent call last):
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/kubecf/dev/linters/run_yamllint.py", line 2, in <module>
    from yamllint import cli
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/yamllint_pypi__yamllint_1_19_0/yamllint/cli.py", line 25, in <module>
    from yamllint import linter
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/yamllint_pypi__yamllint_1_19_0/yamllint/linter.py", line 19, in <module>
    import yaml
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/yamllint_pypi__PyYAML_5_1_2/yaml/__init__.py", line 374
    class YAMLObject(metaclass=YAMLObjectMetaclass):
                              ^
SyntaxError: invalid syntax
```

This appears to be because:

```
$ cat bazel-bin/dev/linters/yamllint
#!/usr/bin/env python

…
```

This appears to be https://github.com/bazelbuild/bazel/issues/4815. We seem to need to change how bazel is run to fix this:

```suggestion
bazel run --incompatible_use_python_toolchains //dev/linters:yamllint -- \
```

Looking around, maybe my version of bazel (0.26) is too old?  Ah, yes, installing bazel 0.29 (the newest version in `obs://devel:tools:building`) fixed things.  All good, then.
